emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteWhitelistHashManager.java

```diff
@@ -0,0 +1,291 @@
+                        "roblox_potion_dance", "kazotsky_kick", "twerk", "club_penguin_dance"
+                    };
+                    for (String emoteName : internalEmotes) {
+                        String jsonPath = "/assets/" + CommonData.MOD_ID + "/emotes/" + emoteName + ".json";
+                        try (InputStream emoteStream = EmoteWhitelistHashManager.class.getResourceAsStream(jsonPath)) {
+                            if (emoteStream != null) {
+                                Path targetFile = whitelistDir.resolve(emoteName + ".json");
+                                Files.copy(emoteStream, targetFile);
+                                copiedEmotes.add(emoteName);
+                            } else {
+                                CommonData.LOGGER.warn("Internal emote resource not found: {}", jsonPath);
+                            }
+                        }
+                    }
+                    if (!copiedEmotes.isEmpty()) {
+                        CommonData.LOGGER.info("Added internal emotes to whitelist directory: {}", String.join(", ", copiedEmotes));
+                    }
+                } catch (Exception e) {
+                    CommonData.LOGGER.warn("Failed to copy internal emotes to whitelist directory", e);
+                }
+            }
+        } catch (IOException e) {
+            CommonData.LOGGER.warn("Failed to create whitelist emotes directory: {}", whitelistDir.toAbsolutePath(), e);
+        }
+    }
+
+    public static class EmoteFileInfo {
+        public String fileName;
+        public long lastModified;
+        public int hash;
+        public EmoteFileInfo(String fileName, long lastModified, int hash) {
+            this.fileName = fileName;
+            this.lastModified = lastModified;
+            this.hash = hash;
+        }
+    }
+
+    public static EmoteWhitelistHashManager getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new EmoteWhitelistHashManager(InstanceService.INSTANCE.getGameDirectory().resolve(Serializer.getConfig().whitelistedEmotesDir.get()));
+        }
+        return INSTANCE;
+    }
+
+    public EmoteWhitelistHashManager(Path whitelistDir) {
+        this.whitelistDir = whitelistDir;
+        this.hashesFile = whitelistDir.resolve(HASHES_FILE);
+    }
+
+    public boolean isHashAllowed(int hash) {
+        return allowedHashes.contains(hash);
+    }
+
+    /**
+     * Force refresh: delete emoteHashes.json and rebuild from whitelist folder
+     */
+    public void forceRefreshHashes() {
+        CommonData.LOGGER.warn("Refreshing all emotes for the whitelist [first run or updated Emotecraft], this may take a while if you have many whitelisted emotes.");
+        try {
+            if (Files.exists(hashesFile)) {
+                Files.delete(hashesFile);
+            }
+        } catch (Exception e) {
+            CommonData.LOGGER.warn("Failed to delete emoteHashes.json for force refresh", e);
+        }
+    }
```

Comment:
Do not save the hashes to a file, just load them with the server load
You can make a separate executor for it, and block the server thread after loading if the hashes have not had time to load